src/cargo/ops/cargo_output_metadata.rs

```diff
@@ -0,0 +1,185 @@
+use std::borrow::Cow;
+use std::collections::HashMap;
+use std::io::Write;
+use std::fs;
+use std::path::{Path, PathBuf};
+use rustc_serialize::{json, Decoder, Decodable};
+use toml;
+use util::CargoResult;
+use util::config::Config;
+use term::color::BLACK;
+use core::{Source, Package, PackageId};
+use core::manifest::{Target};
+use core::registry::PackageRegistry;
+use core::resolver::{Resolve, Method};
+use sources::PathSource;
+use ops;
+
+#[derive(RustcDecodable)]
+pub enum OutputFormat {
+	Toml,
+	Json,
+}
+
+/// Where the dependencies should be written to.
+pub enum OutputTo {
+	Path(PathBuf),
+	StdOut,
+}
+
+impl Decodable for OutputTo {
+	fn decode<D: Decoder>(d: &mut D) -> Result<OutputTo, D::Error> {
+		d.read_option(|d, b| {
+			if b {
+				Ok(OutputTo::Path(try!(Decodable::decode(d))))
+			} else {
+				Ok(OutputTo::StdOut)
+			}
+		})
+	}
+}
+
+pub struct OutputMetadataOptions<'a> {
+	pub output_to: OutputTo,
+	pub output_format: OutputFormat,
+    pub manifest_path: &'a Path,
+    pub features: Vec<String>,
+}
+
+/// Loads the manifest, resolves the dependencies of the project to the concrete
+/// used versions - considering overrides - and writes all dependencies in a TOML
+/// format to stdout or the specified file.
+///
+/// The TOML format is e.g.:
+/// ```toml
+/// root = "libA"
+///
+/// [packages.libA]
+/// dependencies = ["libB"]
+/// path = "/home/user/.cargo/registry/src/github.com-1ecc6299db9ec823/libA-0.1"
+/// version = "0.1"
+///
+/// [packages.libB]
+/// dependencies = []
+/// path = "/home/user/.cargo/registry/src/github.com-1ecc6299db9ec823/libB-0.4"
+/// version = "0.4"
+///
+/// [packages.libB.features]
+/// featureA = ["featureB"]
+/// featureB = []
+/// ```
+pub fn output_metadata(opt: OutputMetadataOptions, config: &Config) -> CargoResult<()> {
+    let (resolved_deps, packages) =
+        try!(resolve_dependencies(opt.manifest_path, opt.features, config));
+
+	#[derive(RustcEncodable)]
+    struct PackageInfo<'a> {
+        version: String,
+        path: Cow<'a, str>,
+        dependencies: Vec<String>,
+        features: Option<&'a HashMap<String, Vec<String>>>,
+        targets: &'a[Target]
+    };
```

Comment:
I think that this may want to avoid redefining custom structures here as it may accidentally be losing information. For example the `dependencies` array as an array of strings will lead to ambiguities where packages have the same name but different versions. I think it may be better to implement `RustcEncodable` on the primitives themselves (e.g. `SourceId` or `PackageId`) as that way it can ensure that no information is being lost and all information is being emitted.